src/esssans/direct_beam.py

```diff
@@ -99,44 +98,44 @@ def direct_beam(pipeline: Pipeline, I0: sc.Variable, niter: int = 5) -> List[dic
     direct_beam_function = None
     bands = pipeline.compute(ProcessedWavelengthBands)
     band_dim = (set(bands.dims) - {'wavelength'}).pop()
-
     full_wavelength_range = sc.concat([bands.min(), bands.max()], dim='wavelength')
 
     pipeline = pipeline.copy()
-    # Append full wavelength range as extra band. This allows for running only a
-    # single pipeline to compute both the I(Q) in bands and the I(Q) for the full
-    # wavelength range.
-    pipeline[ProcessedWavelengthBands] = sc.concat(
-        [bands, full_wavelength_range], dim=band_dim
-    )
-
-    results = []
 
-    # Compute checkpoints to avoid recomputing the same things in every iteration
-    checkpoints = (
-        TransmissionFraction[SampleRun],
-        TransmissionFraction[BackgroundRun],
-        SolidAngle[SampleRun],
-        SolidAngle[BackgroundRun],
-        CleanMonitor[SampleRun, Incident],
-        CleanMonitor[BackgroundRun, Incident],
+    wavelength_bins = pipeline.compute(WavelengthBins)
+    parts = (
         CleanSummedQ[SampleRun, Numerator],
+        CleanSummedQ[SampleRun, Denominator],
         CleanSummedQ[BackgroundRun, Numerator],
+        CleanSummedQ[BackgroundRun, Denominator],
     )
+    parts = pipeline.compute(parts)
+    # Convert events to histograms to make normalization (in every iteration) cheap
+    for key in [
+        CleanSummedQ[SampleRun, Numerator],
+        CleanSummedQ[BackgroundRun, Numerator],
+    ]:
+        parts[key] = parts[key].hist(wavelength=wavelength_bins)
```

Comment:
Should we check here that `parts[key].bins is not None`?

Reply:
Or maybe we leave it because we want an error to be raised if non-event data is being fed through so that we are not silently doing things we don't expect?

Reply:
Yes, should we leave it until we get a requirement to make this work for histogrammed data?